### ISSUE TITLE
Interrupt query?

### ISSUE THREAD

**Issue by zwrmueller** [sqlitebrowser/sqlitebrowser#422]
Hello,

is there any way to Interrupt a (far too) long running query. Actually I kill the process, but that's not the nice way to do.

Regards
 Ralf

**Comment by justinclift** [comment]
Good question.  I don't think there's a way yet, but it would make sense to add one. :smile:

**Comment by MKleusberg** [comment]
No, there is no way to interrupt a query. And unfortunately it's not that easy to fix because of the way SQLite works. As far as I can see we'd have to put the query execution into a separate thread which we can kill from the UI thread at any time - but this requires too much code restructuring to have it happen soon. I'm also not sure if this can somehow cause data loss in the database, so I'm not sure if it's wise to add an easy option which might encourage users to cause unwanted damages.

So in short, I'm not sure on this. But let's keep the issue open as a reminder and maybe come back to it later :smiley:

**Comment by zwrmueller** [comment]
Hello Martin,
I understand your concerns. Perhaps just as information: SQLite Studio shows up an interrupt button when a query is engaged, works fine, so far no problems, but not heavily used anyway.

I came to this, because by un(for)seen changes at my collected measurement data I run into troubles with my (old) queries: SQLite Studio could interrupt them, SQLite Browser I had to kill each time ...

Personally I much more like SQLite Browser; I use SQLite Studio for some work because here I can attach a bunch of databases by a set of query statements, which SQLite Browser cannot.


Regards
 Ralf

**Comment by mglatz** [comment]
please add this :) Not an SQL expert but it should be fairly safe to execute query as a transaction in separate thread and revert the transaction on thread interruption...?

**Comment by MKleusberg** [comment]
@zwrmueller @mglatz  Sorry for the long delay here. Are you still interested in this issue? I'm asking because we are currently preparing some changes which add, besides making some other changes, a cancel button. Because it's changing a lot of our code we want to test this as good as we can. Are you interested in giving it a try? You can find the Windows and OSX build here:

* https://nightlies.sqlitebrowser.org/win32-onceoffs/DB%20Browser%20for%20SQLite-sqlcipher-issue1575v2-win32.msi
* https://nightlies.sqlitebrowser.org/win64-onceoffs/DB%20Browser%20for%20SQLite-sqlcipher-issue1575v2-win64.msi
* https://nightlies.sqlitebrowser.org/osx-onceoffs/DB%20Browser%20for%20SQLite_issue-1575v1.dmg
* https://nightlies.sqlitebrowser.org/osx-onceoffs/DB%20Browser%20for%20SQLite-sqlcipher_issue-1575v1.dmg

**Comment by mglatz** [comment]
I was really hoping to see this in the alpha released recently. It seems to work quite well.

however switching to other tab while query is executing (eg the pragmas tab) displays the "query is runnging" message and there the app seems to freeze

**Comment by mglatz** [comment]
I did some more testing and its not as flawless as i thoguht

- Db browser still freezes on queries with large (15M records)  tables. It seems that cancelling the query within first few seconds work but once it runs for 10 or more seconds, it cannot be cancelled anymore
- running select stement un-greys (enables) the write and revert changes buttons (it should not do that)
- calcelling the query leaves the result window in random states. Once I see some "loading" records, other time there are some values and some "loading" etc. Once the query is cancelled the results should be cleared.

**Comment by justinclift** [comment]
Yep, definitely some weirdness that needs fixing. :wink:

**Comment by sky5walk** [comment]
Confused why the [existing SQLite interrupt api](https://www.sqlite.org/c3ref/interrupt.html), "void sqlite3_interrupt(sqlite3*);" means "a lot of code changes"?

> It is safe to call this routine from a thread different from the thread that is currently running the database operation.

**Comment by chrisjlocke** [comment]
> Confused why the existing SQLite interrupt api, "void sqlite3_interrupt(sqlite3*);" means "a lot of code changes"?

Think of DB4S as a car.  SQLite is the engine.  The engine has a sensor that tells you whether its running or not, and even how many revs its doing.  You devise a convoluted system (no offense...) to take that sensor's output and display it to the user in some rev counter using LEDs.  Pretty.  Works OK.  But someone else now wants it connected to a rotary dial display.  "Its only a sensor.  Should be easy".  However, all the plumbing, design and wiring is geared towards getting the output of that sensor to power LEDs.  Its going to have be ripped out and redesigned to work for an analogue needle, rather than digital LEDs.
Or something.

Basically, the code in DB4S isn't geared up that way, and to make it so, would require a lot of code changes.  We're exceeding 300 issues as it is, with a finite list of programmers.  Its a lot of work.  If we had teams of programmers on standby bored waiting for input, then it could be done....

**Comment by sky5walk** [comment]
I mean no disrespect, as I use DB4S sometimes daily. 😄 
The car analogy is fine. The interrupt api = turning off the ignition. The car's systems are presented with no power and must act accordingly.
It is already mentioned that users(and me 👿) have to use TaskMgr to kill hung DB4S processes.
Can that be any worse than DB4S attempting a sqlite3_interrupt() from a single gui button?

**Comment by justinclift** [comment]
> Can that be any worse than DB4S attempting a sqlite3_interrupt() from a single gui button?

That sums things up pretty well. :smile:

**Comment by MKleusberg** [comment]
Closing this issue as it looks like it has been fixed a while ago and we didn't hear anything since then. Feel free to reopen this issue if the problem still occurs.
